# Курсовая работа № 2
## Аналитика. Начальный уровень









Перед вами стоит задача – подготовить аналитический отчет для HR-отдела. На основании проведенной аналитики предполагается составить рекомендации для отдела кадров по стратегии набора персонала, а также по взаимодействию с уже имеющимися сотрудниками.
<br><br> В базе данных лежит набор таблиц, которые содержат данные о сотрудниках вымышленной компании.
Сделайте обзор штата сотрудников компании. Составьте набор предметов исследования, а затем проверьте их на данных. Вся аналитика должна быть выполена с помощью SQL. Впоследствии данные можно визуализировать, однако финальные датафреймы для графиков также должны быть подготовлены с помощью SQL. <br><br>

Примеры гипотез:
1. Есть зависимость между `perfomance score` и тем, под чьим руководством работает сотрудник.
2. Есть зависимость между продолжительностью работы в компании и семейным положением сотрудника.
2. Есть зависимость между продолжительностью работы в компании и возрастом сотрудника.

<br><br>
Параметры для подключения следующие: хост – `dsstudents.skillbox.ru`, порт – `5432`, имя базы данных – `human_resources`, пользователь – `readonly`, пароль – `6hajV34RTQfmxhS`. Таблицы, доступные для анализа, – `hr_dataset`, `production_staff`, `recruiting_costs`, `salary_grid`.

In [2]:
pip install psycopg2-binary

Note: you may need to restart the kernel to use updated packages.


In [2]:
# импорт необходимых библиотек
import sqlalchemy
import psycopg2
import pandas as pd

# для подключения к БД используем параметры для подключения
conn = 'postgresql+psycopg2://readonly:6hajV34RTQfmxhS@dsstudents.skillbox.ru:5432/human_resources'

engine = sqlalchemy.create_engine(conn)
connect = engine.connect()

print('Первая гипотеза: зависимость между причинами окончания карьеры в компании и количеством увольнений')
# для просмотра первой таблицы и назначение одной из гипотез

hr_dataset = pd.read_sql_query('''select * from  hr_dataset;''', connect)
hr_dataset.head()

print(pd.read_sql_query('''select *from hr_dataset;''', conn).columns)


pd.read_sql_query('''select "Reason For Term",
                                count(*) as amount, round(avg(cast("Days Employed" as numeric)),0) as mean_work_days from hr_dataset where ("Reason For Term" <> 'N/A - still employed')
                                 and ("Reason For Term" <> 'N/A - Has not started yet')
                                group by "Reason For Term"
                                order by "amount" desc;''', connect)

Первая гипотеза: зависимость между причинами окончания карьеры в компании и количеством увольнений
Index(['id', 'Employee Name', 'Employee Number', 'marriedid',
       'maritalstatusid', 'genderid', 'empstatus_id', 'deptid', 'perf_scoreid',
       'age', 'Pay Rate', 'state', 'zip', 'dob', 'sex', 'maritaldesc',
       'citizendesc', 'Hispanic/Latino', 'racedesc', 'Date of Hire',
       'Days Employed', 'Date of Termination', 'Reason For Term',
       'Employment Status', 'department', 'position', 'Manager Name',
       'Employee Source', 'Performance Score'],
      dtype='object')


Reason For Term  amount  mean_work_days
0                   Another position      20           760.0
1                            unhappy      14           743.0
2                         more money      11           762.0
3                      career change       9           766.0
4                              hours       9           505.0
5                         attendance       7          1122.0
6             relocation out of area       5          1002.0
7                   return to school       5           352.0
8                        performance       4           538.0
9                           military       4          1014.0
10                          retiring       4          1148.0
11                  no-call, no-show       3            14.0
12                    medical issues       3           724.0
13  maternity leave - did not return       3          1387.0
14                  gross misconduct       1          1596.0

Вывод: можно сказать что люди, в основном, увольняются по этим причинам: другая должность,несчастье, нехватка денег, смена карьеры. Рекомендации HR-отделу по подборке персонала: спрашивать при приеме на работу насколько претенденту будет нравиться будущая должность, каковы денежные ожидания и рост заработной платы в будущем, обговорить  обязанности работника и их возможный карьерный рост в компании. Если говорить о средней продолжительности работы в компании, то можно сказать, что сотрудники, которые уволились, работали в среднем около 760 рабочих дней, если брать топ 3 причины увольнений.

In [74]:
# для подключения к БД используем параметры для подключения
conn = 'postgresql+psycopg2://readonly:6hajV34RTQfmxhS@dsstudents.skillbox.ru:5432/human_resources'

engine = sqlalchemy.create_engine(conn)
connect = engine.connect()

print('Вторая гипотеза: зависимость между отделом и источником поиска работника. Рассмотрим самые 3 топ отдела,где большое количество сотрудников')
# для просмотра первой таблицы и назначение одной из гипотез

hr_dataset = pd.read_sql_query('''select * from  hr_dataset;''', connect)
hr_dataset.head()

IT_IS_df = pd.read_sql_query('''select "department","Employee Source", count(*) as amount from hr_dataset 
                                where ("Employment Status" = 'Active') and "department" in ('IT/IS')
                                group by "department","Employee Source"
                                order by amount desc;''', connect)
print(IT_IS_df.head())

Sales_df = pd.read_sql_query('''select "department","Employee Source", count(*) as amount from hr_dataset 
                                where ("Employment Status" = 'Active') and "department" in ('Sales')
                                group by "department","Employee Source"
                                order by amount desc;''', connect)
print(Sales_df.head())

Production_df = pd.read_sql_query('''select "department","Employee Source", count(*) as amount from hr_dataset 
                                where ("Employment Status" = 'Active') and "department" in ('Production       ')
                                group by "department","Employee Source"
                                order by amount desc;''', connect)
print(Production_df.head()) 

Вторая гипотеза: зависимость между отделом и источником поиска работника. Рассмотрим самые 3 топ отдела,где большое количество сотрудников
  department       Employee Source  amount
0      IT/IS     Employee Referral      10
1      IT/IS                Indeed       8
2      IT/IS  Professional Society       4
3      IT/IS             Glassdoor       3
4      IT/IS    Diversity Job Fair       3
  department         Employee Source  amount
0      Sales  Pay Per Click - Google       6
1      Sales      Website Banner Ads       6
2      Sales             Monster.com       3
3      Sales      Diversity Job Fair       2
4      Sales                   Other       2
          department       Employee Source  amount
0  Production            Employee Referral      12
1  Production         Professional Society      11
2  Production           Newspager/Magazine      11
3  Production                     MBTA ads      10
4  Production                    Billboard       9


Вывод: рассмотрим 3 отдела: IT/IS, Sales, Producation, в котором работают наибольшее количество сотрудников и определим с помощью каких ресурсов были найдены работники. Если просмотреть эти три отдела, то можно сделать выводы, что источник сотрудника в основном - это Employee Referral и Professional Society  для двух отделов: Production и IT/IS. Что касаемо отдела Sales, для поиска людей в этот отдел можно не использовать этот источник, так как он не очень популярен, но стоит присмотреться к Pay Per Click - Google и Website Banner Ads. Также большую роль в поиске сотрудников играет интернет.

In [124]:
conn = 'postgresql+psycopg2://readonly:6hajV34RTQfmxhS@dsstudents.skillbox.ru:5432/human_resources'

engine = sqlalchemy.create_engine(conn)
connect = engine.connect()

print('Третья гипотеза: зависимость между руководителем и процент сотрудников с удовлетворительной оценкой')
# для просмотра первой таблицы и назначение одной из гипотез

hr_dataset = pd.read_sql_query('''select * from  hr_dataset;''', connect)
hr_dataset
df = pd.read_sql_query('''select name, (CAST(amount_fully AS float)/CAST(amount AS float)*100) as percent from (select pd1.*, pd2.* from (select "Manager Name",
                                count("Performance Score") as amount from hr_dataset 
                                group by "Manager Name"
                                order by "Manager Name") as pd1 
                                join (select "Manager Name"as name, count("Performance Score") as amount_fully from hr_dataset
                                where "Performance Score" = 'Fully Meets'
                                group by "Manager Name"
                                order by "Manager Name") as pd2
                                on pd1."Manager Name" = pd2."name") as dp
                                where amount > 10
                                order by percent desc
                                ;''', connect)


df

# # если посмотреть на руководителей, у которых количество сотрудников меньше 6, то у них более 60% сотрудников имеют оценку Fully Meets 
# # поэтому поставим диапазон выше 10 оценок всего у каждого руководителя и оценим результаты



Третья гипотеза: зависимость между руководителем и процент сотрудников с удовлетворительной оценкой


name    percent
0       John Smith  78.571429
1    David Stanley  71.428571
2       Janet King  68.421053
3    Kelley Spirea  68.181818
4    Ketsia Liebig  66.666667
5   Kissy Sullivan  63.636364
6    Lynn Daneault  61.538462
7     Elijiah Gray  59.090909
8   Webster Butler  52.380952
9         Amy Dunn  52.380952
10    Peter Monroe  42.857143
11  Michael Albert  40.909091
12  Brannon Miller  38.095238
13      Simon Roup  35.294118

Вывод: Можно сказать, что хуже всего справляются с обязанностями руководителя, те работники, у которых наблюдается процент Perfomance score = Fully meets, меньше 60% от общего количества, стоит пересмотреть этих руководителей, как управленческий персонал или проверить отдел HR, который занимается подбором персонала к этим руководителям.